pennylane/__init__.py

```diff
@@ -250,31 +261,77 @@ def grad(func, argnum=None):
         function with respect to the differentiable arguments, or, if specified,
         the arguments in ``argnum``.
     """
-    # pylint: disable=no-value-for-parameter
-    if argnum is not None:
-        # for backwards compatibility with existing code
-        # that manually specifies argnum
-        return _grad(func, argnum)
 
-    def _gradient_function(*args, **kwargs):
-        """Inspect the arguments for differentiability, and
-        compute the autograd gradient function with required argnums
-        dynamically.
+    def __init__(self, fun, argnum=None, *args, **kwargs):
+        self._forward = None
+        self._grad_fn = None
 
-        This wrapper function is returned to the user instead of autograd.grad,
-        so that we can take into account cases where the user computes the
-        gradient function once, but then calls it with arguments that change
-        in differentiability.
+        self._fun = fun
+        self._argnum = argnum
+        self._grad_args = args
+        self._grad_kwargs = kwargs
+
+        if self._argnum is not None:
+            # If the differentiable argnum is provided, we can construct
+            # the gradient function at once during initialization
+            self._grad_fn = grad._grad_with_forward(fun, argnum=argnum, *args, **kwargs)
```

Comment:
More out of curiosity since it is a static method, but why `grad._grad_with_forward` and not `self._grad_with_forward`?

Reply:
Ah, this is just convention. There isn't really much difference between the two different ways of calling a static method.
```suggestion
            self._grad_fn = self._grad_with_forward(fun, argnum=argnum, *args, **kwargs)
```